pkg/sql/systemprivilege/README.md

```diff
@@ -0,0 +1,394 @@
+	USAGE                Kind = 9
+	ZONECONFIG           Kind = 10
+	CONNECT              Kind = 11
+	RULE                 Kind = 12
+	MODIFYCLUSTERSETTING Kind = 13
+)
+```
+
+Once that is done, create a list of the valid privileges for your object type.
+```go
+// Predefined sets of privileges.
+var (
+	AllPrivileges    = List{ALL, CONNECT, CREATE, DROP, SELECT, INSERT, DELETE, UPDATE, USAGE, ZONECONFIG}
+	ReadData         = List{SELECT}
+	ReadWriteData    = List{SELECT, INSERT, DELETE, UPDATE}
+	DBPrivileges     = List{ALL, CONNECT, CREATE, DROP, ZONECONFIG}
+	TablePrivileges  = List{ALL, CREATE, DROP, SELECT, INSERT, DELETE, UPDATE, ZONECONFIG}
+	SchemaPrivileges = List{ALL, CREATE, USAGE}
+	TypePrivileges   = List{ALL, USAGE}
+	// SequencePrivileges is appended with TablePrivileges as well. This is because
+	// before v22.2 we treated Sequences the same as Tables. This is to avoid making
+	// certain privileges unavailable after upgrade migration.
+	// Note that "CREATE, INSERT, DELETE, ZONECONFIG" are no-op privileges on sequences.
+	SequencePrivileges = List{ALL, USAGE, SELECT, UPDATE, CREATE, DROP, INSERT, DELETE, ZONECONFIG}
+	SystemPrivileges   = List{ALL, MODIFYCLUSTERSETTING}
+)
+```
+
+Add privilege checks for your new object type, this depends on where you need the checks.
+
+**IMPORTANT** For 22.1, you **must** version gate the privilege check as we need
+to ensure the `system.privileges` table has been added.
+
+See below for an example of privilege checking + version gating.
+
+The example below checks for **either** the `MODIFYCLUSTERSETTING` role option 
+or "global" privilege, we can only check the privilege if we're on at least
+on version `clusterversion.SystemPrivilegesTable`.
+
+```go
+		hasAdmin, err := p.HasAdminRole(ctx)
+		if err != nil {
+			return err
+		}
+		if !hasAdmin {
+			hasModify, err := p.HasRoleOption(ctx, roleoption.MODIFYCLUSTERSETTING)
+			if err != nil {
+				return err
+			}
+			hasView, err := p.HasRoleOption(ctx, roleoption.VIEWCLUSTERSETTING)
+			if err != nil {
+				return err
+			}
+			if !hasModify && !hasView {
+				if p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.SystemPrivilegesTable) {
+					// We check for EITHER the MODIFYCLUSTERSETTING or VIEWCLUSTERSETTING
+					// role option OR the MODIFYCLUSTERSETTING system cluster privilege.
+					// We return the error for "system cluster privilege" due to
+					// the long term goal of moving away from coarse-grained role options.
+					if err := p.CheckPrivilege(ctx, catalog.SystemClusterPrivilegeObject, privilege.MODIFYCLUSTERSETTING); err != nil {
+						return err
+					}
+				} else {
+					return pgerror.Newf(pgcode.InsufficientPrivilege,
+
+						"only users with either %s or %s privileges are allowed to read "+
+							"crdb_internal.cluster_settings", roleoption.MODIFYCLUSTERSETTING, roleoption.VIEWCLUSTERSETTING)
+				}
+			}
+		}
+```
+
+#### Testing
+Finally, add testing. This can be done using logic tests.
```

Comment:
can this also mention testing the syntax in pkg/sql/parser/testdata ?

Reply:
Done